Extend compaction_history table with additional compaction statistics

paszkow opened the pull request:
Currently, the `system.compaction_history` table miss information like the type of compaction (cleanup, major, resharding, etc), the sstable generations involved (in and out), shard's id the compaction was triggered on and statistics on purged tombstones to be collected during compaction.

The series extends the table with the following columns:

-  "compaction_type" (text)
- "shard_id" (int)
- "sstables_in" (list<sstableinfo_type>)
- "sstables_out" (list<sstableinfo_type>)
- "total_tombstone_purge_attempt" (long)
- "total_tombstone_purge_failure_due_to_overlapping_with_memtable" (long)
- "total_tombstone_purge_failure_due_to_overlapping_with_uncompacting_sstable" (long)

with a user defined type `sstableinfo_type` that holds the information about sstable file

- generation (uuid)
- origin (text)
- size (long)

Additional statistics stored in the compaction_history have been incorporated in the API  `/compaction_manager/compaction_history` and the `nodetool compactionhistory` command.

No backport is required. It extends the existing compaction history output.

Fixes https://github.com/scylladb/scylladb/issues/3791

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12786/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12786/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
#### Failed Tests (8/426):
- [test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12786/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables.debug.1)
- [test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12786/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables.debug.2)
- [test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12786/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables.debug.4)
- [test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12786/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables_debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables.debug.7)
- [rest_api.test_compactionhistory.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12786/testReport/junit/%28root%29/non-boost%20tests/rest_api_test_compactionhistory_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+rest_api.test_compactionhistory.debug.1)
- [rest_api.test_compactionhistory.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12786/testReport/junit/%28root%29/non-boost%20tests/rest_api_test_compactionhistory_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+rest_api.test_compactionhistory.debug.2)
- [rest_api.test_compactionhistory.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12786/testReport/junit/%28root%29/non-boost%20tests/rest_api_test_compactionhistory_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+rest_api.test_compactionhistory.debug.4)
- [rest_api.test_compactionhistory.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12786/testReport/junit/%28root%29/non-boost%20tests/rest_api_test_compactionhistory_debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+rest_api.test_compactionhistory.debug.7)
#### Build Details:
- Duration: 1 hr 42 min
- Builder: i-0897ac2deb34b5ea2 (m5ad.8xlarge)

mykaul commented:
Is a feature flag needed for this?

paszkow commented:
> Is a feature flag needed for this?

I think yes but only for a special case when a user decides to abort upgrading a cluster for some reasons. If a node was upgraded and new data was added to the new columns, then there is a problem with downgrading a node. It will succeed but reading from the compaction_history table fails.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12864/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12864/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
#### Failed Tests (2/486):
- [rest_api.test_compactionhistory.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12864/testReport/junit/%28root%29/non-boost%20tests/rest_api_test_compactionhistory_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+rest_api.test_compactionhistory.release.3)
- [test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12864/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_tombstone_purge_statistics_overlapping_with_other_sstables.release.3)
#### Build Details:
- Duration: 1 hr 53 min
- Builder: i-0ae2e4104c589d416 (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12900/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12900/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/41/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12567/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1470/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4541/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12900/)
#### Failed Tests (13/41406):
- [test_sstablelevelreset](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/offline_tools_test/TestOfflineTools/Tests___dtest_with_tablets___test_sstablelevelreset) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sstablelevelreset)
- [test_sstablelevelreset](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/offline_tools_test/TestOfflineTools/Tests___dtest___test_sstablelevelreset) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sstablelevelreset)
- [test_sstablelevelreset](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/offline_tools_test/TestOfflineTools/Tests___dtest_with_topology_changes___test_sstablelevelreset) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sstablelevelreset)
- [test_replace_with_background_workload[use_endpoint-rbo_enabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest_with_topology_changes___test_replace_with_background_workload_use_endpoint_rbo_enabled_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_background_workload[use_endpoint-rbo_enabled])
- [test_nested_type_dropping](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/user_types_test/TestUserTypes/Tests___dtest___test_nested_type_dropping) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_nested_type_dropping)
- [test_nested_type_dropping](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/user_types_test/TestUserTypes/Tests___dtest_with_tablets___test_nested_type_dropping) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_nested_type_dropping)
- [test_nested_type_dropping](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/user_types_test/TestUserTypes/Tests___dtest_with_topology_changes___test_nested_type_dropping) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_nested_type_dropping)
- [test_type_dropping](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/user_types_test/TestUserTypes/Tests___dtest_with_tablets___test_type_dropping) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_type_dropping)
- [test_type_dropping](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/user_types_test/TestUserTypes/Tests___dtest___test_type_dropping) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_type_dropping)
- [test_type_dropping](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/user_types_test/TestUserTypes/Tests___dtest_with_topology_changes___test_type_dropping) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_type_dropping)
#### Build Details:
- Duration: 4 hr 44 min
- Builder: spider4.cloudius-systems.com

paszkow commented:
> * [test_sstablelevelreset](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/offline_tools_test/TestOfflineTools/Tests___dtest_with_tablets___test_sstablelevelreset) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sstablelevelreset)
> * [test_sstablelevelreset](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/offline_tools_test/TestOfflineTools/Tests___dtest___test_sstablelevelreset) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sstablelevelreset)
> * [test_sstablelevelreset](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/offline_tools_test/TestOfflineTools/Tests___dtest_with_topology_changes___test_sstablelevelreset) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sstablelevelreset)

The entire test file is now completely removed, see the commit [ 576d6d8f05d415ae0d22419d6ef70f4cf7fb2853](https://github.com/scylladb/scylla-dtest/pull/5205/commits/576d6d8f05d415ae0d22419d6ef70f4cf7fb2853)

> * [test_replace_with_background_workload[use_endpoint-rbo_enabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest_with_topology_changes___test_replace_with_background_workload_use_endpoint_rbo_enabled_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_background_workload%5Buse_endpoint-rbo_enabled%5D)

This failed test is irrelevant to the change. Double checked though and it passes locally.

> * [test_nested_type_dropping](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/user_types_test/TestUserTypes/Tests___dtest___test_nested_type_dropping) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_nested_type_dropping)
> * [test_nested_type_dropping](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/user_types_test/TestUserTypes/Tests___dtest_with_tablets___test_nested_type_dropping) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_nested_type_dropping)
> * [test_nested_type_dropping](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/user_types_test/TestUserTypes/Tests___dtest_with_topology_changes___test_nested_type_dropping) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_nested_type_dropping)
> * [test_type_dropping](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/user_types_test/TestUserTypes/Tests___dtest_with_tablets___test_type_dropping) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_type_dropping)
> * [test_type_dropping](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/user_types_test/TestUserTypes/Tests___dtest___test_type_dropping) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_type_dropping)
> * [test_type_dropping](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12900/testReport/junit/user_types_test/TestUserTypes/Tests___dtest_with_topology_changes___test_type_dropping) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_type_dropping)

The above tests will be fixed with the following PR [https://github.com/scylladb/scylla-dtest/pull/5267](https://github.com/scylladb/scylla-dtest/pull/5267)

denesb commented on db/system_keyspace.cc:
```diff
@@ -2322,6 +2322,8 @@ future<> system_keyspace::make(
         co_await db.find_column_family(table).init_storage();
     }
+
+    db.find_keyspace(NAME).add_user_type(sstableinfo_type);
```
Does this get persisted to the disk, into the schema tables?

paszkow replied:
The call is required to have the UDT defined in `system_schema.types` table. Otherwise it is not there.

denesb commented on db/system_keyspace.cc:
```diff
@@ -52,6 +52,21 @@
 
+template <>
+inline sstables::basic_info&& value_cast<sstables::basic_info>(data_value&& value) {
+    auto* user_type = dynamic_cast<const user_type_impl*>(&value._type->without_reversed());
```
Why the `without_reversed()`? Non clustering-key types will never be reversed.

denesb commented on db/system_keyspace.cc:
```diff
@@ -52,6 +52,21 @@
 using days = std::chrono::duration<int, std::ratio<24 * 3600>>;
 
+template <>
+inline sstables::basic_info&& value_cast<sstables::basic_info>(data_value&& value) {
```
Where is this method used?

paszkow replied:
It is indirectly used here https://github.com/scylladb/scylladb/pull/21288/files#diff-5bcbb172f03dae93b568d237d4761d4b1bb01ae8ea23c3ed4279153aa18a5f61R2482.

denesb commented on db/system_keyspace.cc:
```diff
@@ -52,6 +52,21 @@
+    auto* user_type = dynamic_cast<const user_type_impl*>(&value._type->without_reversed());
+    auto* field_values = (const user_type_impl::native_type*)value._value;
+
+    static thread_local sstables::basic_info native_value; // Static so it survives return from this function.
```
why does it need to survive return?

denesb replied:
Ah, I see you return `rvalue` ref, but why?

paszkow replied:
Due to https://github.com/paszkow/scylladb/blob/issue-3791/types/types.hh#L513. See the specialization for `timeuuid_native_type` a few lines below as well.

denesb replied:
But then this is not a cast anymore. It is a conversion.

paszkow replied:
Dropped.

denesb commented on compaction/compaction.cc:
```diff
@@ -892,7 +892,15 @@ class compaction {
+            .shard_id = this_shard_id(),
+            .type = _type,
+            .sstables_in = boost::copy_range<std::vector<sstables::basic_info>>(_sstables | boost::adaptors::transformed(sstable_to_basic_info)),
+            .sstables_out = boost::copy_range<std::vector<sstables::basic_info>>(_all_new_sstables | boost::adaptors::transformed(sstable_to_basic_info)),
```
Please use std ranges.

raphaelsc replied:
I think we may have to compute sstables_in upfront (in setup() probably), due to incremental compaction which may remove input sstables earlier (and this must be preserved, as we cannot keep reference to the sstables after they're removed to allow disk space to be released).

paszkow replied:
Changed into std ranges and moved extracting relevant information from input sstables into setup() function.

denesb commented on compaction/compaction.cc:
```diff
@@ -892,7 +892,15 @@ class compaction {
     }
 protected:
     virtual compaction_result finish(std::chrono::time_point<db_clock> started_at, std::chrono::time_point<db_clock> ended_at) {
+        static auto sstable_to_basic_info = [] (auto sst) {
```
No need for the static IMO, just add a ctor to the struct instead.

denesb commented on tools/scylla-nodetool.cc:
```diff
@@ -109,6 +109,38 @@ struct fmt::formatter<file_size_printer> : fmt::formatter<string_view> {
+    int64_t total_tombstone_purge_attempt;
+    int64_t total_tombstone_purge_failure_due_to_overlapping_with_memtable;
+    int64_t total_tombstone_purge_failure_due_to_overlapping_with_uncompacting_sstable;
+};
```
Can't we reuse the type used in Scylla core? You can include any header here.

paszkow replied:
Reused

denesb reviewed (changes requested):
@raphaelsc please also review.

raphaelsc commented:
> > Is a feature flag needed for this?
> 
> I think yes but only for a special case when a user decides to abort upgrading a cluster for some reasons. If a node was upgraded and new data was added to the new columns, then there is a problem with downgrading a node. It will succeed but reading from the compaction_history table fails.

Isn't it better to add compaction_history_v2? That's how we've been doing it. I am a little worried about adding new feature flags for every schema change. We add another field tomorrow, and then we need another feature flag, and it becomes a maintainance burden. Also I believe it's not cheap either to keep adding new feature flags since each one adds some overhead to communication. It's not the end of the world if compaction history is lost. It's essentially persisted logging.

raphaelsc commented:
I chatted with Botond, and we indeed don't want to cause rollback headaches, so please disregard my comment about simply doing the schema changes without any handling. I am fine with the feature flag. I admit I am concerned about accumulation of feature flags, but we can remove them down the road.

paszkow commented:
In v2:
- Rebase
- Drop `value_cast<sstables::basic_info>` implementation in favor of a deserialization function that transforms data_value into sstables::basic_info.
- Extract compaction_history_entry struct into a seperate files and re-use it in scylla-nodetool
- Correctly adjust` nodetool/test_compactionhistory.py` tests

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13710/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13710/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/588/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5251/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2145/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13270/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13710/)
#### Failed Tests (1/40996):
- [test_decommission_not_enough_racks.release.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13710/testReport/junit/topology_experimental_raft/test_tablets/Tests___Unit_Tests___test_decommission_not_enough_racks_release_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_not_enough_racks.release.6)
#### Build Details:
- Duration: 4 hr 43 min
- Builder: spider2.cloudius-systems.com

mykaul commented:
CI failure is probably unrelated - and it's https://github.com/scylladb/scylla-enterprise/issues/5060

denesb commented on db/system_keyspace_compaction_history_entry.hh:
```diff
@@ -0,0 +1,28 @@
+
+namespace db {
+
+struct system_keyspace_compaction_history_entry {
```
I think this name is too verbose, both for the file and for the struct. Let's leave it as just `compaction_history_entry`.

paszkow replied:
Done in v3

denesb commented on db/system_keyspace.cc:
```diff
@@ -606,8 +606,15 @@ schema_ptr system_keyspace::built_indexes() {
             {"bytes_out", long_type},
```
The commit message has to be updated, the offset thing is no longer required, system-schema versions are now hash based and are automatically updated when the table changes.

paszkow replied:
Ah, true. Forgot to change it.

denesb commented on types/types.cc:
```diff
@@ -3465,6 +3465,8 @@ thread_local const shared_ptr<const abstract_type> varint_type(make_shared<varin
 thread_local const shared_ptr<const abstract_type> counter_type(make_shared<counter_type_impl>());
 thread_local const shared_ptr<const abstract_type> duration_type(make_shared<duration_type_impl>());
+thread_local const shared_ptr<const user_type_impl> sstableinfo_type(user_type_impl::get_instance(
+        "system", "sstableinfo", {"generation", "origin", "size"}, {uuid_type, utf8_type, long_type}, false));
```
These are our CQL types, lets not add new types here.
You can add this type to `db/system_keyspace.cc` instead. See also https://github.com/scylladb/scylladb/pull/21088 which also added a UDT to a system table.

paszkow replied:
Done in v3

denesb commented on test/nodetool/test_compactionhistory.py:
```diff
@@ -150,6 +270,26 @@ def test_yaml(nodetool):
+                        }],
+                    "total_tombstone_purge_attempt": 14,
+                    "total_tombstone_purge_failure_due_to_overlapping_with_memtable": 0,
+                    "total_tombstone_purge_failure_due_to_overlapping_with_uncompacting_sstable": 0,
```
Please also update the nodetool docs.

paszkow replied:
Done in v3

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13885/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13885/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:x: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/138/)
:x: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/736/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2338/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5457/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13469/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13885/)
#### Failed Tests (1/41111):
- [test_index_doesnt_flood_cache_in_small_partition_workload](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13885/testReport/junit/boost/cache_algorithm_test/test_index_doesnt_flood_cache_in_small_partition_workload) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_index_doesnt_flood_cache_in_small_partition_workload)
#### Build Details:
- Duration: 4 hr 36 min
- Builder: spider4.cloudius-systems.com

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/21288)

**Changed Files:**
- [operating-scylla/nodetool-commands/compactionhistory](http://previews.docs.scylladb.com/scylladb/21288/operating-scylla/nodetool-commands/compactionhistory)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

paszkow commented:
In v3:

- Moved away a custom user type from types.hh
- Add missing updates to nodetool documentation
- Renamed extracted compaction history structure to `compaction_history_entry`

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13910/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13910/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
#### Failed Tests (2/986):
- [rest_api.test_compactionhistory.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13910/testReport/junit/%28root%29/non-boost%20tests/rest_api_test_compactionhistory_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+rest_api.test_compactionhistory.debug.1)
- [test_compactionhistory_rows_merged_time_window_compaction_strategy.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13910/testReport/junit/rest_api/test_compactionhistory/test_compactionhistory_rows_merged_time_window_compaction_strategy_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_rows_merged_time_window_compaction_strategy.debug.1)
#### Build Details:
- Duration: 1 hr 55 min
- Builder: i-09c8543bac8768bba (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13909/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13909/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/756/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5483/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/158/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13495/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2366/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13909/)
#### Failed Tests (1/41119):
- [test_mv_topology_change.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13909/testReport/junit/topology_custom/test_mv_topology_change/Tests___Unit_Tests___test_mv_topology_change_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_topology_change.dev.3)
#### Build Details:
- Duration: 8 hr 59 min
- Builder: i-01a56617a7ad3b6ae (m5ad.8xlarge)

denesb commented on db/compaction_history_entry.hh:
```diff
@@ -0,0 +1,28 @@
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
```
License should be our sorce-available license.

denesb commented on db/system_keyspace.cc:
```diff
@@ -53,6 +53,9 @@
 
 using days = std::chrono::duration<int, std::ratio<24 * 3600>>;
```
commit message still has the prefix: types.

denesb commented on mutation/mutation_tombstone_stats.hh:
```diff
@@ -0,0 +1,31 @@
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
```
License.

denesb reviewed (changes requested):
Code LGTM. Please update the licenses to our new source-available one.

@raphaelsc please review.

paszkow commented:
In v4:

- Rebase, and fix commit message and licence

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14134/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14134/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/928/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2551/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13677/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/332/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5664/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14134/)
#### Build Details:
- Duration: 4 hr 56 min
- Builder: spider4.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14135/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14135/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/930/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2554/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5667/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13680/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/334/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14135/)
#### Build Details:
- Duration: 9 hr 28 min
- Builder: spider4.cloudius-systems.com

paszkow commented:
> Code LGTM. Please update the licenses to our new source-available one.
> 
> @raphaelsc please review.

@raphaelsc Ping. Please review.

mykaul commented:
@paszkow - this has missed 2025.1.x, and has conflicts.

paszkow commented:
> @paszkow - this has missed 2025.1.x, and has conflicts.

It missed because I received only one review. Since we all focused on tablets-related issues, it had lower priority. I will resolve the conflicts.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15825/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15825/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3859/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/15002/)
:x: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6973/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15825/)
#### Build Details:
- Duration: 8 hr 8 min
- Builder: spider2.cloudius-systems.com

avikivity commented:
random idea: add a switch to `nodetool compactionhistrory` to make its output be like a standard log. Then we can merge its output with regular log output (when needed) and remove INFO level compaction logging. @denesb

avikivity commented on compaction/compaction_manager.cc:
```diff
@@ -385,7 +385,7 @@ future<sstables::compaction_result> compaction_task_executor::compact_sstables_a
+        co_await update_history(*_compacting_table, std::move(res), cdata);
     }
 
     co_return res;
```
How can you move res if you co_return it here?

How come nothing broke?

paszkow replied:
That is a mistake. The most interesting is, however, the second question. Turns out that I move `res` when `should_update_history` is true and in this particular case, the caller ignores the returned result. So it was just a luck.

avikivity commented on db/compaction_history_entry.hh:
```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
```
2025

avikivity commented on db/compaction_history_entry.hh:
```diff
@@ -3,26 +3,37 @@
 
 /*
- * SPDX-License-Identifier: AGPL-3.0-or-later
+ * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
```
This should have been in the patch that introduced the new file, not here.

paszkow replied:
Fixed. Moved to the correct patch.

raphaelsc commented:
> random idea: add a switch to `nodetool compactionhistrory` to make its output be like a standard log. Then we can merge its output with regular log output (when needed) and remove INFO level compaction logging. @denesb

I suggested exactly this somewhere, I think it would be valuable. Let's make sure the changes in this PR paves the way for it.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -614,8 +614,15 @@ schema_ptr system_keyspace::built_indexes() {
+            {"sstables_out", list_type_impl::get_instance(sstableinfo_type, false)},
+            {"total_tombstone_purge_attempt", long_type},
+            {"total_tombstone_purge_failure_due_to_overlapping_with_memtable", long_type},
+            {"total_tombstone_purge_failure_due_to_overlapping_with_uncompacting_sstable", long_type},
```
These aren't conditional.

(we should have a helper that allows building schemas where some columns are tied to a feature&)

paszkow replied:
Why should they be? 

I think that so far we have been adding columns unconditionally. The feature flag is used to ensure no data is added to the new columns until all nodes are upgraded. As long as nothing is written there, a rollback is possible.

avikivity commented on compaction/compaction_garbage_collector.hh:
```diff
@@ -22,7 +22,19 @@ using can_gc_fn = std::function<bool(tombstone, is_shadowable)>;
+    std::optional<timestamp_source> source;
+};
+
+using max_purgeable_fn = std::function<max_purgeable(const dht::decorated_key&, is_shadowable)>;
```
@denesb could this have an impact on the read path?

raphaelsc replied:
We calculate max purgeable once per partition for a reason. I am not sure we're increasing number of times here, but if we do, it's certainly a regression.

denesb replied:
This changes the return value from bool to some struct. I don't think will have any visible impact.
Like Raphael said, we call this method once per each partition (maybe twice, if there are shadowable tombstones in the partition too).

paszkow replied:
I do not increase the frequency the function is called. Only the return type is changed.

avikivity replied:
Is the complexity of calculating the return type materially increased? If not, ok.

raphaelsc commented:
> > random idea: add a switch to `nodetool compactionhistrory` to make its output be like a standard log. Then we can merge its output with regular log output (when needed) and remove INFO level compaction logging. @denesb
> 
> I suggested exactly this somewhere, I think it would be valuable. Let's make sure the changes in this PR paves the way for it.

For example, compaction_history has a timepoint where compaction ended (compacted_at), but not when it started. It's needed for achieving that. @paszkow please address this.

avikivity commented on mutation/mutation_compactor.hh:
```diff
@@ -182,6 +183,7 @@ class compact_mutation_state {
     std::unique_ptr<mutation_compactor_garbage_collector> _collector;
 
     compaction_stats _stats;
+    tombstone_purge_stats* _tombstone_stats = nullptr;
```
It may be better to make this a reference and update it unconditionally than to add new branches. This is what we usually do.

I say may because it looks like the statistics collection is complicated, but let's see.

paszkow replied:
I considered using a reference in the first place. I need to, however, pass the statistics structure from the `compaction` class through the `compacting_reader` or `compact_for_compaction_v2` to the `compact_mutation_state`.

That would require creating dummy structures in places where intermediate functions are called. That's why I went with the pointer.

avikivity commented on mutation/mutation_compactor.hh:
```diff
@@ -250,14 +252,36 @@ private:
+                    } else if (timestamp_source.value() == max_purgeable::timestamp_source::other_sstables_possibly_shadowing_data) {
+                        ++_tombstone_stats->failures_due_to_overlapping_with_uncompacting_sstable;
+                    }
+                }
```
I have an allergy to stats code using branches. How about making the counters part of an array, and using the timestamp_source as an index into that array?

avikivity commented on mutation/mutation_compactor.hh:
```diff
@@ -293,6 +317,7 @@ private:
             _max_purgeable = _get_max_purgeable(*_dk, is_shadowable);
         }
         auto ret = t.timestamp < _max_purgeable.timestamp;
+        timestamp_source = _max_purgeable.source;
```
Please avoid output parameters, they reduce readability at the call site. Return a struct.

avikivity replied:
Also, perhaps we can drop the optional by adding an enumeration option for none.

paszkow replied:
Done

avikivity commented on mutation/mutation_tombstone_stats.hh:
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
```
2025

avikivity commented:
Looks good overall.

paszkow commented:
I a new version:

- Fix licencing errors
- Added `started_at` column
- Added `None` enum value in `max_purgeable::timestamp_source`
- Use `None` as default timestamp source in `max_purgeable` struct instead of optional
- Use the `timestamp_source` as an index into an array to collect stats instead of adding branches
- Replace the output parameter in `compact_mutation_state::can_gc` and return a pair instead

paszkow commented on sstables/open_info.hh:
```diff
@@ -41,6 +41,12 @@ struct entry_descriptor {
+    sstring origin;
+    int64_t size;
+};
+
```
Moved to a separarate file.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16567/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16567/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16567/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/7692/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/15737/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4551/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16567/)
#### Failed Tests (4/45267):
- [test_replace_with_same_ip_twice.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16567/testReport/junit/cluster/test_replace_with_same_ip_twice/Tests___Unit_Tests___test_replace_with_same_ip_twice_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_same_ip_twice.debug.3)
- [test_replace_with_same_ip_twice.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16567/testReport/junit/cluster/test_replace_with_same_ip_twice/Tests___Unit_Tests___test_replace_with_same_ip_twice_debug_3_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_same_ip_twice.debug.3)
- [test_replace_with_same_ip_twice.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16567/testReport/junit/cluster/test_replace_with_same_ip_twice/Tests___Unit_Tests___test_replace_with_same_ip_twice_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_same_ip_twice.dev.2)
- [test_replace_with_same_ip_twice.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16567/testReport/junit/cluster/test_replace_with_same_ip_twice/Tests___Unit_Tests___test_replace_with_same_ip_twice_dev_2_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_same_ip_twice.dev.2)
#### Build Details:
- Duration: 6 hr 49 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16568/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16568/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16568/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4557/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/15750/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/7702/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16568/)
#### Failed Tests (2/45234):
- [test_replace_with_same_ip_twice.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16568/testReport/junit/cluster/test_replace_with_same_ip_twice/Tests___Unit_Tests___test_replace_with_same_ip_twice_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_same_ip_twice.debug.1)
- [test_replace_with_same_ip_twice.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16568/testReport/junit/cluster/test_replace_with_same_ip_twice/Tests___Unit_Tests___test_replace_with_same_ip_twice_debug_1_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_same_ip_twice.debug.1)
#### Build Details:
- Duration: 10 hr
- Builder: spider9.cloudius-systems.com

paszkow commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16568/) ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16568/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16568/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/mutation_test 🔹 boost/sstable_compaction_test 🔹 nodetool/test_compactionhistory 🔹 rest_api/test_compactionhistory ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4557/) ✅ - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/15750/) ✅ - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/7702/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16568/)
> 
> #### Failed Tests (2/45234):
> * [test_replace_with_same_ip_twice.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16568/testReport/junit/cluster/test_replace_with_same_ip_twice/Tests___Unit_Tests___test_replace_with_same_ip_twice_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_same_ip_twice.debug.1)
> * [test_replace_with_same_ip_twice.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16568/testReport/junit/cluster/test_replace_with_same_ip_twice/Tests___Unit_Tests___test_replace_with_same_ip_twice_debug_1_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_same_ip_twice.debug.1)
> 
> #### Build Details:
> * Duration: 10 hr
> * Builder: spider9.cloudius-systems.com

https://github.com/scylladb/scylladb/issues/23553

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17264/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17264/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17264/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5202/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16452/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8369/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17264/)
#### Failed Tests (1/46127):
- [test_cache_compacts_expired_tombstones_on_read](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17264/testReport/junit/boost/row_cache_test/test_cache_compacts_expired_tombstones_on_read) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cache_compacts_expired_tombstones_on_read)
#### Build Details:
- Duration: 8 hr 36 min
- Builder: spider3.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17308/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17308/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17308/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5241/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8412/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16495/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17308/)
#### Build Details:
- Duration: 6 hr 38 min
- Builder: spider5.cloudius-systems.com

bhalevy commented:
@denesb / @scylladb/scylla-maint looks like the PR is ready to be merged.

bhalevy commented:
@paszkow it looks like 27933692882f10c9546f1ac9609811ece3c7c943 would be helpful in the field when we suspect tombstone accumulation.  How easy would it be to backport it?  Do we need other patches from this series?

Cc @raphaelsc
